### PR TITLE
API Fetch: Normalize keys of incoming preloaded data

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to resolve an issue where preloaded data using `createPreloadingMiddleware` makes assumptions about the paths provided in the preloaded data, which if not met, can lead to unintentional cache misses when later API fetches attempt to reference the preloaded data.

**Example:**

```js
createPreloadingMiddleware( {
    'wp/v2/demo?foo=bar&baz=quux': { body: { /* ... */ } }
} );

// ...

apiFetch( 'wp/v2/demo?baz=quux&foo=bar' );
```

Prior to these changes, the `apiFetch` call above would not use the preloaded data.

The preloading middleware internally normalizes paths to reference identical query parameters as the same, regardless of their original order. However, this normalization was not applied to incoming preloaded data. Thus, since the preloaded data above assigns keys in the non-normalized form, they would never be matched.

**Testing Instructions:**

Verify unit tests pass:

```
npm run test-unit packages/api-fetch/src/middlewares/test/preloading.js
```

### CODE REVIEW

**Review by youknowriad — APPROVED** [review]
LGTM but I wonder if this means we should push a Core patch to remove the normalization that is already happening on the backend (because I think there's something there)

**Comment by youknowriad** [comment]
At least, I do remember some normalization happening on the backend, maybe it's not the case anymore.

**Comment by aduth** [comment]
I don't think there's anything special happening on the backend.

I think it's just this hard-coded list:

https://github.com/WordPress/wordpress-develop/blob/29f0d74/src/wp-admin/edit-form-blocks.php#L42-L53

(There's no default filtering on `block_editor_preload_paths` and [`rest_preload_api_request`](https://github.com/WordPress/wordpress-develop/blob/29f0d7459978c139f3e2323570f72297e89b9069/src/wp-includes/rest-api.php#L1497-L1563) doesn't modify the path in any way)

I'm actually curious though, looking at the default list, whether there are existing issues, since the normalization should be alphabetizing the order of the query parameters (thus, the preloading for `/taxonomies` and `/users/me` may not be matching?)

**Comment by aduth** [comment]
> I'm actually curious though, looking at the default list, whether there are existing issues, since the normalization should be alphabetizing the order of the query parameters (thus, the preloading for `/taxonomies` and `/users/me` may not be matching?)

It would appear this is the case. When loading the editor, I see a request for the `/taxonomies` and `/users/me` endpoints in the Chrome Network tab.

For one thing, I think we changed away from requesting `per_page=-1`, and instead use [`fetchAllMiddleware`](https://github.com/WordPress/gutenberg/blob/master/packages/api-fetch/src/middlewares/fetch-all-middleware.js) to request 100 taxonomies at a time. I expect we probably want to change this in core to just preload the first page of data.

**Comment by aduth** [comment]
> For one thing, I think we changed away from requesting `per_page=-1`, and instead use [`fetchAllMiddleware`](https://github.com/WordPress/gutenberg/blob/master/packages/api-fetch/src/middlewares/fetch-all-middleware.js) to request 100 taxonomies at a time. I expect we probably want to change this in core to just preload the first page of data.

While this is true, the way it works is that a component will express their query as if they're fetching `per_page: -1`, and the middleware transforms the path to request 100 at a time. But for the purposes of preloading, it considers the path as `per_page=-1`.

It's still not matched because the preloaded parameter order is not normalized (fixed by this pull request).

Whether we should change the order of these middlewares so that it tries to use preloaded data for the segment of data actually being fetched (i.e. first page of `per_page=100`), it seems reasonable. It would be something to explore separately (in Gutenberg, not core).

File: packages/api-fetch/src/middlewares/preloading.js [hunk 1]
```diff
@@ -1,49 +1,65 @@
...
+	return base + '?' + query
+		// [ 'b=1', 'c=2', 'a=5' ]
+		.split( '&' )
+		// [ [ 'b, '1' ], [ 'c', '2' ], [ 'a', '5' ] ]
+		.map( function( entry ) {
+			return entry.split( '=' );
+		} )
+		// [ [ 'a', '5' ], [ 'b, '1' ], [ 'c', '2' ] ]
+		.sort( function( a, b ) {
+			return a[ 0 ].localeCompare( b[ 0 ] );
```
[the remark this answers is not in the document]
Comment (aduth) [reply packages/api-fetch/src/middlewares/preloading.js]: > TIL `localeCompare`. Nifty. How will this perform if the string values happen to be different locales though. Is that a concern here? I'm guessing no, but just thought I'd surface it.

Even if it differed, as long as it's consistent within a locale, it should work just as well.

Also noting, this wasn't added here. The pull request merely shuffled it around, and included some tests which were previously lacking.

**Pull request merged**